File: problems/041-progressive-cost-budget/solutions/cpp/main.cpp
#include <cstdint>
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::size_t n, q;
    std::cin >> n >> q;
    std::vector<std::uint64_t> costs(n);
    for (auto &cost : costs) {
        std::cin >> cost;
    }

    std::size_t completed = 0;
    std::uint64_t spent = 0;
    for (std::size_t query = 0; query < q; ++query) {
        std::uint64_t budget;
        std::cin >> budget;
        while (completed < n && costs[completed] <= budget - spent) {
            spent += costs[completed++];
        }
        std::cout << completed << '\n';
    }
}

File: problems/041-progressive-cost-budget/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReaderSize(os.Stdin, 1<<20)
	out := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	costs := make([]uint64, n)
	for index := range costs {
		fmt.Fscan(in, &costs[index])
	}

	completed := 0
	var spent uint64
	for query := 0; query < q; query++ {
		var budget uint64
		fmt.Fscan(in, &budget)
		for completed < n && costs[completed] <= budget-spent {
			spent += costs[completed]
			completed++
		}
		fmt.Fprintln(out, completed)
	}
}
